### PR TITLE
v4.0.x: Fix bug where --timeout won't kill children.

### PR DESCRIPTION
**Pull request by awlauria**
Add a new state to kill the procs when --timeout is hit.

bot:notacherrypick

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

### CODE REVIEW

**Comment by awlauria** [comment]
Can potentially replace https://github.com/open-mpi/ompi/pull/9272

**Comment by rhc54** [comment]
I don't get it - you are saying that this also solves your observed problem?? Yet all your new state does is have mpirun kill its local procs and then call `orte_quit`, which also kills the local procs and then stops the event loop.

The "all_jobs_complete" state the code currently uses just goes to `orte_quit`, which kills the local procs and then stops the event loop.

So how can these two states not generate the exact same behavior?

**Comment by awlauria** [comment]
> I don't get it - you are saying that this also solves your observed problem??

Yes.

> Yet all your new state does is have mpirun kill its local procs and then call `orte_quit`, which also kills the local procs and then stops the event loop.

Does orte_quit call killprocs? I don't see this in the code. I did not add a call to killprocs, I only added a state that would get us in this if block.

> The "all_jobs_complete" state the code currently uses just goes to `orte_quit`, which kills the local procs and then stops the event loop.

I do not see this, I see killprocs only called in two functions, `job_errors()`, and `proc_errors()`.

The current behavior, without this patch, would set the orted state to: `ORTE_JOB_STATE_ALL_JOBS_COMPLETE`, and not call killprocs() like I would expect.

File: orte/orted/orted_submit.c [hunk 1]
```diff
@@ -3226,7 +3226,7 @@ void orte_timeout_wakeup(int sd, short args, void *cbdata)
     }
   giveup:
     /* abort the job */
-    ORTE_ACTIVATE_JOB_STATE(NULL, ORTE_JOB_STATE_ALL_JOBS_COMPLETE);
+    ORTE_ACTIVATE_JOB_STATE(NULL, ORTE_PROC_STATE_TIMED_OUT);
```
Comment (awlauria) [inline orte/orted/orted_submit.c:3229]: Hm, this is the wrong error-code type.

**Comment by awlauria** [comment]
So I see two solutions: 1

changing the state set on --timeout to: `ORTE_JOB_STATE_COMM_FAILED` from `ORTE_JOB_STATE_ALL_JOBS_COMPLETE`. This will initiate all procs being killed in the `kill_jobs()` switch in `orte/mca/errmgr/default_orted/errmgr_default_orted.c`

or two : add a ORTE_JOB_STATE_TIMED_OUT and add it to the switch in `kill_jobs()` to kill the procs. Basically the same solution presented here - but re-done at the job level and using the proper error code.

**Comment by rhc54** [comment]
> changing the state set on --timeout to: ORTE_JOB_STATE_COMM_FAILED from ORTE_JOB_STATE_ALL_JOBS_COMPLETE. This will initiate all procs being killed in the kill_jobs() switch in orte/mca/errmgr/default_orted/errmgr_default_orted.c

No it doesn't - you are looking at the wrong errmgr component. Only the HNP sees the timeout code, and if you look at the HNP component, when you pass a NULL job (as timeout does) all we do (regardless of state) is execute the "FORCED_EXIT" state - which literally calls "terminate_orteds" and exits.

> or two : add a ORTE_JOB_STATE_TIMED_OUT and add it to the switch in kill_jobs() to kill the procs. Basically the same solution presented here - but re-done at the job level and using the proper error code.

Yeah, the proc state confused me. See above - it does the same thing.

In both cases, all you've done is use the state machine to do what the other PR added - a call to terminate the orteds. Using the state machine is the correct approach. However, it still begs the question of why the orteds aren't properly cleaning up their child procs when suiciding.

I'm trying to think of some way to see the orted state machine trace once mpirun times out and goes away. The ssh connection will be gone, so you can't get output that way. Might require telling opal_output to dump to a file so you can see it. I think that's the only way to get to the bottom of this - we need to understand the difference timeout is making. I'm beginning to wonder if that `shutdown` call is the key and simply needs to be removed.

**Comment by awlauria** [comment]
From what I can tell, sending a kill command to mpirun (which we think should have the same behavior as --timeout) goes down the following path:

```
$ pkill mpirun
$ [[44867,0],0] ACTIVATE JOB NULL STATE DAEMONS TERMINATED AT orted/orted_comm.c:382
```
which is done under the case `ORTE_DAEMON_EXIT_CMD`, which in turn calls:

```
        /* kill the local procs */
        orte_odls.kill_local_procs(NULL);
```

which is why I don't see left-over processes when I send `mpirun` a sigterm at the command-line.

It seems to me that the sigterm path goes down the kill procs path - but not the --timeout path with the current states.

**Comment by rhc54** [comment]
I'm afraid you missed some critical code path there. The ess/hnp component sets up a trap for SIGTERM that flows into the `clean_abort` routine - which calls `terminate_orteds`. It is that call which winds up in orted_comm killing the local procs, and also sends a clean termination signal to the orteds (which winds up in orted_comm). It then has `mpirun` go thru its normal shutdown procedure.

So that isn't triggering the suicide scenario.

Timeout, on the other hand, is handled in orted/orted_submit.c. If you aren't looking for stack traces, then all it does is activate "ALL_JOBS_COMPLETE", which just goes to orte_quit, kills the local procs, and then stops the event loop. This lets `mpirun` fall thru its normal shutdown procedure (but doesn't send a message to the orteds to terminate). Close to the suicide scenario as the only indication the daemons receive for HNP departure is lost communication - with one key difference. When mpirun just dies, the socket simply closes. When mpirun goes thru its normal shutdown procedure, it cleanly closes the OOB, which results in each socket having `shutdown` called on it prior to closing.

If you want to trigger the suicide scenario, you have to hit `mpirun` with SIGKILL so it doesn't get the chance to tell the orteds to die and it doesn't go thru normal shutdown. If you look at the timeout code in orted_submit.c, you'll see that there is an option for testing suicide response (governed by the `ORTE_TEST_HNP_SUICIDE` envar). This was included as a means of testing system response to actual `mpirun` failure. You can use that instead of issuing the SIGKILL if you like.

If we see the orteds doing the right thing under the suicide scenario, then we should remove `shutdown` from the `CLOSE_THE_SOCKET` macro and see if timeout now works correctly. Could be that this one-line deletion is all that is actually required to solve this problem.

**Comment by awlauria** [comment]
@rhc54 removing the `shutdown` call from CLOSE_THE_SOCKET did not work for me. Again - I am not seeing the processes being sent a kill signal from orted as when I send mpirun a SIGTERM. The latter (sending mpirun a sigterm directly) still works as expected in killing the child ranks.

This PR, the and the solutions I outlined here:

```
changing the state set on --timeout to: ORTE_JOB_STATE_COMM_FAILED from ORTE_JOB_STATE_ALL_JOBS_COMPLETE. This will initiate all procs being killed in the kill_jobs() switch in orte/mca/errmgr/default_orted/errmgr_default_orted.c

or two : add a ORTE_JOB_STATE_TIMED_OUT and add it to the switch in kill_jobs() to kill the procs. Basically the same solution presented here - but re-done at the job level and using the proper error code.
```

both work.

**Comment by awlauria** [comment]
> Timeout, on the other hand, is handled in orted/orted_submit.c. If you aren't looking for stack traces, then all it does is activate "ALL_JOBS_COMPLETE", which just goes to orte_quit, kills the local procs, and then stops the event loop. 

I'm not seeing where setting ALL_JOBS_COMPLETE kills jobs. I only see `ORTE_JOB_STATE_COMM_FAILED` results in ranks being killed, which is why I am advocating for an additional state to handle this timeout case. If this is wrong, can you point me the logic that ALL_JOBS_COMPLETE will kill the child processes? Running with `odls_base_verbose 100` I am not seeing ` odls:kill_local_proc` activated.

**Comment by awlauria** [comment]
All this said, I am fine with moving forward with https://github.com/open-mpi/ompi/pull/9272 and being done with it. Both fix the --timeout issue from my perspective. I just submitted this as an alternative patch.

**Pull request closed without merging**